module1/homework/grayscaleImages/compression.cpp

```diff
@@ -0,0 +1,72 @@
+#include "compression.hpp"
+
+#include <algorithm>
+#include <cctype>
+#include <iostream>
+#include <iterator>
+
+std::vector<std::pair<uint8_t, uint8_t>> compressGrayscale(
+    const std::array<std::array<uint8_t, width>, height>& bitmap) {
+    std::vector<std::pair<uint8_t, uint8_t>> compressedBitmap{};
+
+    std::for_each(cbegin(bitmap),
+                  cend(bitmap),
+                  [&compressedBitmap](const auto& row) {
+                      auto iterStart = begin(row);
+                      auto iterFound = iterStart;
+                      while (iterFound != end(row)) {
+                          iterFound = std::find_if(iterStart,
+                                                   end(row),
+                                                   [&iterStart](const auto el) {
+                                                       return el != *iterStart;
+                                                   });
+                          compressedBitmap.push_back({*iterStart, std::distance(iterStart, iterFound)});
+                          iterStart = iterFound;
+                      }
+                  });
+
+    compressedBitmap.shrink_to_fit();
+
+    return compressedBitmap;
+}
+
+std::array<std::array<uint8_t, width>, height> decompressGrayscale(
+    const std::vector<std::pair<uint8_t, uint8_t>>& compressedBitmap) {
+    std::array<std::array<uint8_t, width>, height> decompressedBitmap{};
+
+    auto iterRow = begin(decompressedBitmap);
+    auto iterEl = begin(*iterRow);
+
+    std::for_each(cbegin(compressedBitmap),
+                  cend(compressedBitmap),
+                  [&iterRow, &iterEl](const auto& el) {
+                      iterEl = std::fill_n(iterEl, el.second, el.first);
+                      if (iterEl == end(*iterRow)) {
+                          iterRow++;
+                      }
+                  });
+
+    return decompressedBitmap;
+}
+
+void printMap(const std::array<std::array<uint8_t, width>, height>& bitmap) {
+    constexpr uint8_t space = ' ';
+    constexpr uint8_t newLine = '\n';
+
+    std::transform(std::begin(bitmap),
+                   std::end(bitmap),
+                   std::ostream_iterator<uint8_t>(std::cout),
+                   [](const auto& row) {
+                       std::transform(std::begin(row),
```

Comment:
Here std::copy will be better. You just copy output value to std::ostream. Not transform them.

Reply:
OK, but how in such case handle printing only printable characters? I am using transform to change characters that cannot be printed into space. With std::copy I do not have such possiblity.